File: loopviewpagerlib/src/main/java/mocaris/com/loopviewpagerlib/LoopViewPagerAdapter.java
package mocaris.com.loopviewpagerlib;

import android.content.Context;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;

import java.util.List;

/**
 * 无限轮播 viewPager
 */
public abstract class LoopViewPagerAdapter<T> extends PagerAdapter {

    protected Context context;
    protected ViewPager viewPager;
    protected boolean loop;

    protected List<T> mDatas;

    public LoopViewPagerAdapter(Context context, ViewPager viewPager, boolean loop) {
        this.context = context;
        this.viewPager = viewPager;
        this.loop = loop;
    }

    public List<T> getDatas() {
        return mDatas;
    }

    public void setDatas(List<T> mDatas) {
        this.mDatas = mDatas;
    }

    @Override
    public int getCount() {
        return loop ? Integer.MAX_VALUE : null == mDatas ? 0 : mDatas.size();
    }

    public int getItemCount() {
        return null == mDatas ? 0 : mDatas.size();
    }

    public void loop() {
        if (loop) {
            int count = getCount();
            viewPager.setCurrentItem(count / 2 - (count / 2) % getItemCount());
            viewPager.setOffscreenPageLimit(2);
        }
    }
}

File: demo/src/main/java/mocaris/com/demo/LoopViewPagerAdapter.kt
package mocaris.com.demo

import android.content.Context
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager
import android.view.View
import android.view.ViewGroup

abstract class LoopViewPagerAdapter<T>(val context: Context, val viewPager: ViewPager, val loop: Boolean) : PagerAdapter() {


    lateinit var mDatas: List<T>


    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return view == `object`
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        container.removeView(`object` as View?)
    }

    override fun getCount(): Int {
        return if (loop) Int.MAX_VALUE else mDatas?.size
    }

    fun getitemCount(): Int {
        return mDatas?.size ?: 0
    }

    fun getItem(position: Int): T {
        return mDatas?.get(position % mDatas.size)
    }

    fun loop() {
        if (loop) {
            viewPager?.currentItem = count / 2 - (count / 2) % getitemCount()
            viewPager.offscreenPageLimit = 2
        }
    }
}